File: ios/SpentyAI/SpentyAI/Features/Purchases/PurchaseRepository.swift
import Foundation

// MARK: - Request / Response Models

struct BillPayload: Codable {
    var billNumber: String?
    var vendorId: String?
    var vendorName: String?
    var date: Date?
    var dueDate: Date?
    var lineItems: [BillLineItemPayload]?
    var subtotal: Double?
    var taxAmount: Double?
    var grandTotal: Double?
    var notes: String?
}

struct BillLineItemPayload: Codable {
    var description: String?
    var hsnSac: String?
    var quantity: Double?
    var rate: Double?
    var taxRate: Double?
    var amount: Double?
}

struct RecordBillPaymentPayload: Codable {
    var amount: Double
    var date: Date?
    var paymentMethod: String?
    var accountId: String?
    var notes: String?
}

struct BillNextNumberResponse: Codable {
    let nextNumber: String?
}

struct BillStatsResponse: Codable {
    let total: Int?
    let paid: Int?
    let unpaid: Int?
    let partial: Int?
    let totalBilled: Double?
    let totalPaid: Double?
    let totalOutstanding: Double?
    let totalOverdue: Double?
}

struct BillCountResponse: Codable {
    let count: Int?
}

struct CreditorSummary: Codable, Identifiable {
    var id: String { vendorId ?? UUID().uuidString }
    let vendorId: String?
    let vendorName: String?
    let totalOutstanding: Double?
    let billCount: Int?
    let oldestDate: String?
}

struct AgingBucket: Codable, Identifiable {
    var id: String { label ?? bucket ?? UUID().uuidString }
    let label: String?
    let bucket: String?
    let amount: Double?
    let count: Int?
}

struct VendorPurchaseSummary: Codable, Identifiable {
    var id: String { vendorId ?? UUID().uuidString }
    let vendorId: String?
    let vendorName: String?
    let totalPurchases: Double?
    let billCount: Int?
}

struct BillParseResponse: Codable {
    let vendorName: String?
    let billNumber: String?
    let date: String?
    let dueDate: String?
    let lineItems: [BillLineItemPayload]?
    let subtotal: Double?
    let taxAmount: Double?
    let grandTotal: Double?
    let notes: String?
}

struct BillListResponse: Codable {
    let items: [Bill]
    let total: Int?
}

struct CreditorListResponse: Codable {
    let items: [CreditorSummary]
    let total: Int?
}

struct AgingBucketListResponse: Codable {
    let items: [AgingBucket]?
    let buckets: [String: AgingBucketDetail]?
}

struct AgingBucketDetail: Codable {
    let amount: Double?
    let count: Int?
}

struct VendorPurchaseListResponse: Codable {
    let items: [VendorPurchaseSummary]
    let total: Int?
}

struct AccountListResponse: Codable {
    let accounts: [Account]
}

// MARK: - Repository

final class PurchaseRepository {

    static let shared = PurchaseRepository()
    private init() {}

    private let api = APIClient.shared

    // MARK: - CRUD

    func fetchBills() async throws -> [Bill] {
        let response: BillListResponse = try await api.get(APIEndpoints.bills)
        return response.items
    }

    func fetchBill(id: String) async throws -> Bill {
        try await api.get(APIEndpoints.bill(id))
    }

    func createBill(_ payload: BillPayload) async throws -> Bill {
        try await api.post(APIEndpoints.bills, body: payload)
    }

    func updateBill(id: String, _ payload: BillPayload) async throws -> Bill {
        try await api.put(APIEndpoints.bill(id), body: payload)
    }

    func deleteBill(id: String) async throws {
        let _: MessageResponse = try await api.delete(APIEndpoints.bill(id))
    }

    // MARK: - Actions

    func recordPayment(id: String, _ payload: RecordBillPaymentPayload) async throws -> Bill {
        try await api.post(APIEndpoints.billRecordPayment(id), body: payload)
    }

    func markPaid(id: String) async throws -> Bill {
        let updated: Bill = try await api.post(APIEndpoints.billMarkPaid(id))
        return updated
    }

    func duplicateBill(id: String) async throws -> Bill {
        try await api.post(APIEndpoints.billDuplicate(id))
    }

    // MARK: - Next Number

    func getNextNumber() async throws -> String {
        let response: BillNextNumberResponse = try await api.get(APIEndpoints.billsNextNumber)
        return response.nextNumber ?? ""
    }

    // MARK: - Stats & Analytics

    func fetchStats() async throws -> BillStatsResponse {
        try await api.get(APIEndpoints.billsStats)
    }

    func fetchCreditors() async throws -> [CreditorSummary] {
        let response: CreditorListResponse = try await api.get(APIEndpoints.billsCreditors)
        return response.items
    }

    func fetchAging() async throws -> [AgingBucket] {
        try await api.get(APIEndpoints.billsAging)
    }

    func fetchPurchasesByVendor() async throws -> [VendorPurchaseSummary] {
        let response: VendorPurchaseListResponse = try await api.get(APIEndpoints.billsPurchasesByVendor)
        return response.items
    }

    // MARK: - PDF

    func fetchBillPDF(id: String) async throws -> Data {
        try await api.getRaw(APIEndpoints.billPDF(id))
    }

    // MARK: - Upload & Parse

    func uploadAndParseBill(data: Data, filename: String, mimeType: String) async throws -> BillParseResponse {
        try await api.upload(
            APIEndpoints.billsParseUpload,
            data: data,
            filename: filename,
            mimeType: mimeType
        )
    }

    // MARK: - Vendors & Accounts (for pickers)

    func fetchVendors() async throws -> [Vendor] {
        let response: VendorListResponse = try await api.get(APIEndpoints.vendors)
        return response.items
    }

    func fetchAccounts() async throws -> [Account] {
        let response: AccountListResponse = try await api.get(APIEndpoints.accounts)
        return response.accounts
    }
}

File: android-native/app/src/main/java/com/spentyai/app/features/purchases/PurchaseRepository.kt
package com.spentyai.app.features.purchases

import com.spentyai.app.core.models.Account
import com.spentyai.app.core.models.Bill
import com.spentyai.app.core.models.Vendor
import com.spentyai.app.core.network.ApiClient
import com.spentyai.app.core.network.ApiResult
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.JsonPrimitive

// ---------- Response / Payload models ----------

@Serializable
data class BillStats(
    @SerialName("total_billed") val totalBilled: Double = 0.0,
    @SerialName("total_paid") val totalPaid: Double = 0.0,
    @SerialName("total_outstanding") val totalOutstanding: Double = 0.0,
    @SerialName("total_overdue") val totalOverdue: Double = 0.0
)

@Serializable
data class CreditorSummary(
    @SerialName("vendor_id") val vendorId: String? = null,
    @SerialName("vendor_name") val vendorName: String? = null,
    @SerialName("total_outstanding") val totalOutstanding: Double = 0.0,
    @SerialName("bill_count") val billCount: Int = 0
)

@Serializable
data class BillAgingBucket(
    val label: String? = null,
    val bucket: String? = null,
    val amount: Double = 0.0,
    val count: Int = 0
) {
    val displayLabel: String get() = label ?: bucket ?: "--"
}

// Form line item model
data class PurchaseFormLineItem(
    val id: String = java.util.UUID.randomUUID().toString(),
    var description: String = "",
    var hsnSac: String = "",
    var quantity: Double = 1.0,
    var rate: Double = 0.0,
    var taxRate: Double = 0.0
) {
    val computedAmount: Double get() = quantity * rate
    val taxAmount: Double get() = computedAmount * taxRate / 100.0
    val lineTotal: Double get() = computedAmount + taxAmount
}

class PurchaseRepository(private val apiClient: ApiClient) {

    // CRUD

    suspend fun fetchAll(): ApiResult<List<Bill>> =
        apiClient.safeApiCall { apiClient.endpoints.getBills() }

    suspend fun fetch(id: String): ApiResult<Bill> =
        apiClient.safeApiCall { apiClient.endpoints.getBill(id) }

    suspend fun create(payload: JsonObject): ApiResult<Bill> =
        apiClient.safeApiCall { apiClient.endpoints.createBill(payload) }

    suspend fun update(id: String, payload: JsonObject): ApiResult<Bill> =
        apiClient.safeApiCall { apiClient.endpoints.updateBill(id, payload) }

    suspend fun delete(id: String): ApiResult<JsonObject> =
        apiClient.safeApiCall { apiClient.endpoints.deleteBill(id) }

    // Actions

    suspend fun recordPayment(id: String, payload: JsonObject): ApiResult<Bill> =
        apiClient.safeApiCall { apiClient.endpoints.markBillPaid(id, payload) }

    suspend fun markPaid(id: String): ApiResult<Bill> =
        apiClient.safeApiCall {
            apiClient.endpoints.markBillPaid(id, buildJsonObject { put("mark_fully_paid", true) })
        }

    // Supporting data

    suspend fun fetchVendors(): ApiResult<List<Vendor>> =
        apiClient.safeApiCall { apiClient.endpoints.getVendors() }

    suspend fun fetchAccounts(): ApiResult<List<Account>> =
        apiClient.safeApiCall { apiClient.endpoints.getAccounts() }
}

fun buildBillPayload(
    billNumber: String?,
    vendorId: String?,
    vendorName: String?,
    issueDate: String,
    dueDate: String,
    lineItems: List<PurchaseFormLineItem>,
    subtotal: Double,
    taxAmount: Double,
    grandTotal: Double,
    notes: String?
): JsonObject = buildJsonObject {
    billNumber?.let { put("bill_number", it) }
    vendorId?.let { put("vendor_id", it) }
    vendorName?.let { put("vendor_name", it) }
    put("issue_date", issueDate)
    put("due_date", dueDate)
    put("subtotal", subtotal)
    put("tax_amount", taxAmount)
    put("total", grandTotal)
    notes?.let { put("notes", it) }
    put("line_items", buildJsonArray {
        lineItems.filter { it.description.isNotBlank() }.forEach { item ->
            add(buildJsonObject {
                put("description", item.description)
                if (item.hsnSac.isNotBlank()) put("hsn_sac", item.hsnSac)
                put("quantity", item.quantity)
                put("unit_price", item.rate)
                put("tax_rate", item.taxRate)
                put("amount", item.computedAmount)
            })
        }
    })
}
